src/Orleans.Runtime/Lifecycle/SiloLifetime.cs

```diff
@@ -0,0 +1,251 @@
+                    // Since the stage has already completed, there is no impending timeout for this late registration, so we pass CancellationToken.None.
+                    // For late participants we do not check for termination!
+
+                    await participant.Execute(CancellationToken.None).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "Error executing late-registered callback for silo lifecycle stage = '{StageName}'", name);
+                }
+            }
+        }
+
+        public async Task NotifyCompleted(CancellationToken cancellationToken)
+        {
+            Debug.Assert(!_isNotifyingOrHasCompleted, "This should not be called twice!");
+
+            List<StageParticipant> snapshot;
+
+            lock (_lock)
+            {
+                _isNotifyingOrHasCompleted = true;
+                snapshot = [.. _participants];
+            }
+
+            // If we took the snapshot, no further registrations will be account for the sequential invocations.
+            // Late registrations will still be executed though, just not in order anymore.
+
+            var exceptions = new List<Exception>();
+
+            foreach (var participant in snapshot)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+
+                try
+                {
+                    // TODO: Not sure if we should WaitAsync in case the users ignore the token we pass.
+                    // Or if we should wait indefinatly until the callback executes?!
+
+                    await participant.Execute(cancellationToken).WaitAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    // Ignore, this is expected during hard shutdown.
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "Error executing callback for silo lifecycle = '{StageName}'", name);
+
+                    exceptions.Add(ex);
+
+                    if (participant.TerminateOnError)
+                    { 
+                        // We signal cancellation so any BG-running participants that observe the stage can stop.
+                        CancelStage();
+
+                        // In case there was a previous error but the participant chose to not terminate, we include that error aswell.
+                        var aggException = new AggregateException(exceptions);
+
+                        _tcs.SetException(aggException);
+
+                        throw aggException;
+                    }
+                }
+            }
+
+            CancelStage();
```

Comment:
I think we should do it in this order:

First:
- Run callbacks & `_cts.CancelAsync()`, concurrently, awaiting them via `Task.WhenAll` since there is no ordering implied.

Then:
_ `tcs.SetResult(...)` / `_tcs.SetCanceled(...)` / `_tcs.SetException(...)`

The `CancellationToken` indicates that the stage has started, the `Task` indicates that it has completed, and the callbacks form the body of the stage.